crates/paimon/src/fileindex/file_index_format.rs

```diff
@@ -0,0 +1,486 @@
+// under the License.
+
+use std::{collections::HashMap, io::SeekFrom};
+
+use tokio::io::{AsyncReadExt, AsyncSeekExt, AsyncWrite, AsyncWriteExt, BufReader, BufWriter};
+
+pub const MAGIC: u64 = 1493475289347502;
+pub const EMPTY_INDEX_FLAG: i32 = -1;
+
+/// File index file format. Put all columns and offsets in the header.
+///
+/// ```
+///   _____________________________________    _____________________
+/// ｜     magic    ｜version｜head length  ｜
+/// ｜-------------------------------------｜
+/// ｜            column number            ｜
+/// ｜-------------------------------------｜
+/// ｜   column 1        ｜ index number   ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 1 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 2 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 3 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜            HEAD
+/// ｜   column 2        ｜ index number   ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 1 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 2 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜  index name 3 ｜start pos ｜length  ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜                 ...                 ｜
+/// ｜-------------------------------------｜
+/// ｜  redundant length ｜redundant bytes  ｜
+/// ｜-------------------------------------｜    ---------------------
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜
+/// ｜                BODY                 ｜             BODY
+/// ｜                BODY                 ｜
+/// ｜_____________________________________｜    _____________________
+/// ```
+///
+/// - `magic`: 8 bytes long
+/// - `version`: 4 bytes int
+/// - `head length`: 4 bytes int
+/// - `column number`: 4 bytes int
+/// - `column x`: var bytes utf (length + bytes)
+/// - `index number`: 4 bytes int (how many column items below)
+/// - `index name x`: var bytes utf
+/// - `start pos`: 4 bytes int
+/// - `length`: 4 bytes int
+/// - `redundant length`: 4 bytes int (for compatibility with later versions, in this version, content is zero)
+/// - `redundant bytes`: var bytes (for compatibility with later version, in this version, is empty)
+/// - `BODY`: column index bytes + column index bytes + column index bytes + .......
+
+#[derive(Debug)]
+pub struct FileIndexFormat;
+
+impl FileIndexFormat {
+    pub fn create_writer<W: AsyncWrite + Unpin>(writer: W) -> Writer<W> {
+        Writer::new(writer)
+    }
+
+    pub async fn create_reader<R: AsyncReadExt + AsyncSeekExt + Unpin + Clone>(
+        reader: R,
+    ) -> std::io::Result<Reader<R>> {
+        Reader::new(reader).await
+    }
+}
+
+#[allow(dead_code)]
+pub struct Writer<W: AsyncWrite + Unpin> {
```

Comment:
I feel like we should build our io module first.

- structs like `InputFile` & `OutputFile`
- traits like `FileRead` and `FileWrite`.

We can follow the iceberg pattern: https://github.com/apache/iceberg-rust/blob/main/crates/iceberg/src/io/file_io.rs

Reply:
OK, I will mark this PR as Draft later, and then add the relevant IO to-do items in the corresponding Issue.